Build fails unpredictably if _CoqProject file contains missing a file

#### Version

8.10+alpha(d58614c1adb94fbfbb75e385bb75f768022b9629)

#### Operating system

macOS 10.14.2 & Debian Stretch

#### Description of the problem

E.g. this

```shell
touch a.v
echo 'Require a.' > b.v
printf "a.v\nb.v\n\no_such_file.v" > _CoqProject
coq_makefile -f _CoqProject -o Makefile
make VERBOSE=1
```

produces the following output:

```shell
/Applications/Xcode.app/Contents/Developer/usr/bin/make --no-print-directory -f "Makefile" pre-all
if [ "8.10+alpha" != "8.10+alpha" ]; then\
	  echo "W: This Makefile was generated by Coq 8.10+alpha";\
	  echo "W: while the current Coq version is 8.10+alpha";\
	fi
/Applications/Xcode.app/Contents/Developer/usr/bin/make --no-print-directory -f "Makefile" real-all
make: *** [all] Error 2
```

So it does not provide any pointers as to where to look for the source of error.

Also (this happened on a real project), when there is a missing file (basically a typo in the file name) in `_CoqProject` file, then some of dependencies get resolved incorrectly and the build is failing at an unrelated moment (too early to quickly understand what has caused it).

I can't reproduce it:
```shell
$ coq_makefile foobar.v -o xxx
$ make -f xxx
make[1]: *** No rule to make target 'foobar.v', needed by 'foobar.vo'.  Stop.
xxx:312: recipe for target 'all' failed
```

what is `/Applications/Xcode.app/Contents/Developer/usr/bin/make` ?

Here is another way to reproduce the problem:

```shell
$ git clone https://github.com/imdea-software/fcsl-pcm
$ cd fcsl-pcm
$ echo "foobar.v" >> _CoqProject
$ make
coq_makefile -f _CoqProject -o CoqMakefile
/Applications/Xcode.app/Contents/Developer/usr/bin/make --no-print-directory -f CoqMakefile
COQC finmap/finmap.v
File "./finmap/finmap.v", line 21, characters 25-32:
Error: Unable to locate library ordtype with prefix fcsl.

make[2]: *** [finmap/finmap.vo] Error 1
make[1]: *** [all] Error 2
make: *** [invoke-coqmakefile] Error 2
```

`finmap/finmap.v` depends on `finmap/ordtype.v`, but because of that extra `foobar.v`file  the dependency gets lost and the compilation error is confusing.

--- a/doc/changelog/08-cli-tools/13541-fix9319.rst
+++ b/doc/changelog/08-cli-tools/13541-fix9319.rst
@@ -1,0 +1,5 @@
+- **Fixed:**
+  ``coq_makefile`` has improved logic when dealing with incorrect ``_CoqProject`` files
+  (`#13541 <https://github.com/coq/coq/pull/13541>`_,
+  fixes `#9319 <https://github.com/coq/coq/issues/9319>`_,
+  by Fabian Kunze).

--- a/test-suite/coq-makefile/gen-v-during-make/run.sh
+++ b/test-suite/coq-makefile/gen-v-during-make/run.sh
@@ -1,0 +1,18 @@
+#!/usr/bin/env bash
+
+. ../template/init.sh
+
+cat > _CoqProject <<EOF
+-R ./theories/ ""
+theories/use.v
+theories/generatedInPreAll.v
+EOF
+
+cat > theories/use.v <<EOF
+Require generatedInPreAll.
+EOF
+
+printf 'theories/generatedInPreAll.v:\n\ttouch theories/generatedInPreAll.v\n' > Makefile.local
+
+coq_makefile -f _CoqProject -o Makefile
+make

--- a/test-suite/coq-makefile/missing-included/run.sh
+++ b/test-suite/coq-makefile/missing-included/run.sh
@@ -1,0 +1,21 @@
+#!/usr/bin/env bash
+
+. ../template/init.sh
+
+cat > _CoqProject <<EOF
+-R ./theories/ ""
+theories/b.v
+theories/a.v
+theories/noSuchFile.v
+EOF
+touch theories/a.v
+cat > theories/b.v <<EOF
+Require a.
+EOF
+
+coq_makefile -f _CoqProject -o Makefile
+
+if make 2> stdErr.log; then exit 1; fi
+
+# on osx it says "No rule to make target `.Makefile.d', needed by `theories/b.vo'.  Stop."
+#grep -q noSuchFile stdErr.log

--- a/test-suite/coq-makefile/missing-required/run.sh
+++ b/test-suite/coq-makefile/missing-required/run.sh
@@ -1,0 +1,17 @@
+#!/usr/bin/env bash
+
+. ../template/init.sh
+
+cat > _CoqProject <<EOF
+-R ./theories/ "Lib"
+theories/a.v
+EOF
+cat > theories/a.v <<EOF
+Require noSuchFile.
+EOF
+
+coq_makefile -f _CoqProject -o Makefile
+
+if make 2> stdErr.log; then exit 1; fi
+
+grep -q noSuchFile stdErr.log

--- a/tools/CoqMakefile.in
+++ b/tools/CoqMakefile.in
@@ -759,7 +759,7 @@ else
 TIMING_EXTRA =
 endif
 
-$(VOFILES): %.vo: %.v
+$(VOFILES): %.vo: %.v | $(VDFILE)
 	$(SHOW)COQC $<
 	$(HIDE)$(TIMER) $(COQC) $(COQDEBUG) $(TIMING_ARG) $(COQFLAGS) $(COQLIBS) $< $(TIMING_EXTRA)
 ifeq ($(COQDONATIVE), "yes")

